### ISSUE TITLE
Rule priorities are based on total rule length, not individual path prefix

### ISSUE THREAD

**Issue by drdaeman** [traefik/traefik#1663]
### Do you want to request a feature or report a bug?

A bug. I believe Traefik uses incorrect logic to decide on routing rules priority.

### What did you do?

Consider Traefik is configured to watch for Docker containers (a very basic config that essentially boils down to `[docker] watch = true`), and observes the following configuration in their labels (as a snippet from Docker Compose file):

    web:
        ...
        labels:
            traefik.enable: "true"
            traefik.backend: "web"
            traefik.frontend.rule: "Host:one.example.org,one.example.net,old.one.example.net,old.one.example.org"

    web-ng:
        ...
        labels:
            traefik.enable: "true"
            traefik.backend: "web-ng"
            traefik.frontend.rule: "Host:one.example.org,one.example.net,PathPrefix:/api/v2/,/o/"

### What did you expect to see? What did you see instead?

I've expected a request to `one.example.org/api/v2/` to be routed to the `web-ng` container, as it has more specific `PathPrefix`.

However, requests were sent to `web` instead. This was the case until I've made the `web-ng`'s rule look like `"Host:one.example.org,one.example.net,PathPrefix:/api/,/o/,/a-long-string-to-make-this-rule-more-specific/"`.

I've expected Traefik to compare path prefixes of all matching rules rather than rules themselves. I.e. compare "" (longest PathPrefix that matches from the `web`'s rules - empty in this case) vs "/api/" (longest PathPrefix that matches from the `web-ng`'s rules). I haven't looked at the code but it seems that this isn't the case. I've experimented a little and can tell that traffic is routed to `web-ng` only when the whole rule's length is greater *or equal* to the `web`'s frontend rule's length.

### Output of `traefik version`:

    Version:      v1.3.0-rc3
    Codename:     raclette
    Go version:   go1.8.2
    Built:        2017-05-24_01:39:53PM
    OS/Arch:      linux/amd64

**Comment by timoreimann** [comment]
Looking at the [code](https://github.com/containous/traefik/blob/master/server/server.go#L917), it does really seem like we're using the entire rule content length in order to define the priority.

This does seem wrong. @emilevauge, is this a bug for you too?

**Comment by emilevauge** [comment]
Ouch, indeed, we added multiple rules separated by `,` after priorities... `newRoute.Priority(serverRoute.route.GetPriority() + len(route.Rule))` should definitely be moved to `rules.go`.

**Comment by drdaeman** [comment]
I've tried to create a test that checks for this behavior. Here's what I came with: https://gist.github.com/drdaeman/b37478e2d07207047dd1432229b9e637 (not a PR, because I'm not exactly sure the code's correct)

Unless I've messed up somewhere and misunderstanding things, the final router for two rules, one with `"PathPrefix:/foo,/bar,/foobarbaz"` and another with `"PathPrefix:/foobar"` would still looks like this (very rough schematic representation):

    Router.routes:
      - matchers:
        - Router.routes:
          - matchers:
            - matcher: template: "/foo"
          - matchers:
            - matcher: template: "/bar"
          - matchers:
            - matcher: template: "/foobarbaz"
        - Router.routes.matchers:
          - matchers:
            - matcher: template: "/foobar"

Nested sub`Router`s are created in `Rules.pathPrefix`, so each rule is separate. Whenever a request for "/foobar" comes, it would be caught by the first subrouter and that would still result in an incorrect match with "/foo". Or, depending on priorities/sorting, it could be backwards - request for "/foobarbaz" would match at the wrong subrouter.

As I understand it, the structure must be flattened to work correctly, and subrouters should be created only for host-level rules - with multiple path rules targeting the same host merged together under the single router.

This seems to be a significantly big change :(

**Comment by Fank** [comment]
Same issue here:
1. traefik.frontend.rule: Host:foo.bar,bin.baz
2. traefik.frontend.rule: Host:foo.bar:PathPrefix:/sidebar
3. traefik.frontend.rule: Host:bin.baz:PathPrefix:/ws
4. traefik.frontend.rule: Host:bin.baz:PathPrefix:/v2

Every request will be forwarded to 1.

**Comment by djmaze** [comment]
Just discovered this bug myself and was glad to find this issue. 

In my opinion this odd behaviour is very significant, and as long as the underlying design has not been corrected,  it should be documented (in the official documentation).

**Comment by ldez** [comment]
https://docs.traefik.io/v1.6/basics/#priorities

**Comment by djmaze** [comment]
Oops.. Well, should have RTFM ;) Sorry.

**Comment by aaslamin** [comment]
Hi @jbdoumenjou, I was wondering why you removed the bug label from this issue? 

This doesn't seem like an enhancement to me.

**Comment by SloCompTech** [comment]
I still have and issue have to services linked as:
- `Host("example.org", "www.example.org") && Path("/app")`
- `Host("example.org", "www.example.org")`
Path `www.example.org/app` goes to second instead of first.

**Comment by gabrieldemarmiesse** [comment]
Just my 2c: I spent three hours on this problem. Just by adding a new hostname to a service, the app stopped working. I believe this rule of using the lenght of the  route rule to set priority is against the principle of least surprise. 

Or maybe to make the debugging easier, in the traefik dashboard, the list of routes could be sorted by priority, with the priority being written next to the route name?

**Comment by 1activegeek** [comment]
I'll add in that this seems to add an extra layer of complexity that is really unnecessary. I would suspect the optimal solution would be to use something like the sum of characters that match the rule, with a higher number being a correct priority. Looking at the length of a rule, doesn't necessarily imply a more specific URL match, because there can be simple OR and AND logic added in. Consider:

```
Host(`domainthatisreallylong.org`)
vs
Path(`/path`)
```

While the Host is a longer rule, it's not more specific. In this scenario, it would be assumed that the Host is already inclusive if only hosting 1 domain. If multiple, then someone may need a rule that includes the domain as well to be more specific - or perhaps may want to easily handle the same redirect for multiple domains.

I would propose a weighting mechanism to add MORE weight to something like a path/prefix than host for example. And then additionally beyond that, have each part of the rule that does in fact match, count for say a percentage of the whole match.

So if URL is 20 char, and 10 are of the root domain - this is 50% match, and the other 10 are the path - this is 50%. However in this scenario, a path/prefix would get more weight to imbalance the 1:1 battle of decision. Obviously there is more to this to think through such as a path only being 10% but matching the scenario above would fail. 

The short term fix it seems, is to pick an arbitrary high number to put on rules evaluation much shorter than others in precedence. Or in a scenario where a rule is longer (higher number) set it with an extremely low priority in static config.

**Comment by seuf** [comment]
Is it possible to just display the calculated router priority in the traefik dashboard ?

**Comment by kevtainer** [comment]
> Is it possible to just display the calculated router priority in the traefik dashboard ?

This has been discussed in the past and was declined due to the additional real estate it would take in the UI given the possible range of numbers in the rule (I believe it's a signed 32bit integer, iirc). Calculated priorities also do not seem to be provided via the API, however, manually specified values are.

**Comment by eceglov** [comment]
> > Is it possible to just display the calculated router priority in the traefik dashboard ?
> 
> This has been discussed in the past and was declined due to the additional real estate it would take in the UI given the possible range of numbers in the rule (I believe it's a signed 64bit integer, iirc). Calculated priorities also do not seem to be provided via the API, however, manually specified values are.
> 
> #5981

It would be still useful at least log the priority in the debug mode thus elaborating which exact rule matched for a specific request.

**Comment by seuf** [comment]
I understand it's difficult to display the priority in the table view, but maybe in the router's details section, it should be readable..

**Comment by xavivars** [comment]
Other proxies use # of rules to define priorities. So a matching rule with Host & Path would always have more priority than a rule with only host or only path. That makes it significantly easier to handle priorities without hacks and workarounds.

**Comment by seuf** [comment]
Traefik should also consider adding more priority to more specific rules.
For example :
```
Host(`www.domain.com`) && Path(`/foo`)
```
is more specific than
```
HostRegexp(`{subdomain:.*}.domain.{tld:.*}`) && PathPrefix(`/`)
```
But the second rule is longer so it win ! :disappointed:

**Comment by drdaeman** [comment]
I would also like to mention that there is always an option of introducing an ability to specify rule weight/priority manually, with something in spirit of `Weight(100): Host(`www.example.org`) && Path(`/foo`)`. Although I'm not sure what would be a good syntax for this.

Manual weights are inconvenient when one wants something that "just works" so aren't a complete solution, however they would eliminate any guesswork and allow arbitrarily complex routing policies working without crude hacks (like non-existent long paths that add extra weight).

**Comment by KOPACb** [comment]
Got same issue with path prefixes.
Next work well:
```
        static:
          rule: "PathPrefix(`/`)"
          service: "core-static"
          entryPoints:
            - web
        api:
          rule: "PathPrefix(`/api/`)"
          service: "core"
          entryPoints:
            - web
        microservice-api:
          rule: "PathPrefix(`/microservice/api/`)"
          service: "microservice-api"
          middlewares:
            - "microservice"
          entryPoints:
            - web
        microservice:
          rule: "PathPrefix(`/microservice/`)"
          service: "microservice-static"
          middlewares:
            - "microservice"
          entryPoints:
            - websecure
```

But when i add one more rule:

```
        static:
          rule: "PathPrefix(`/`,  `/api/v2/static`)"
          service: "core-static"
          entryPoints:
            - web
        api:
          rule: "PathPrefix(`/api/`)"
          service: "core"
          entryPoints:
            - web
        microservice-api:
          rule: "PathPrefix(`/microservice/api/`)"
          service: "microservice-api"
          middlewares:
            - "microservice"
          entryPoints:
            - web
        microservice:
          rule: "PathPrefix(`/microservice/`)"
          service: "microservice-static"
          middlewares:
            - "microservice"
          entryPoints:
            - web
```
`/microservice/` starts routing to `/` eg core-static.

**Comment by md-seb** [comment]
This would make so much more sense if it was based on rule specificity, kinda like CSS works.

**Comment by tfny** [comment]
After years of contending with this, we are going to close this issue. 

We just haven't found a way of contending with this that addresses the issue despite approaching it from many perspectives, including using the rule length to set the priorities (splitting, long rules in many routers to lower the priority) and using the [priority](https://doc.traefik.io/traefik/routing/routers/#priority) option.

However we do agree that having the priority in the UI is a good idea, so it will be implemented with [#9004](https://github.com/traefik/traefik/pull/9004) which you can look forward to in Traefik Proxy 3.0.
